File: motion_primitive/motion_primitive_library.py
import numpy as np, time, scipy.optimize
from motion_primitive import MotionPrimitive


class TrajectoryError(Exception):
    def __init__(self, traj):
        self.traj = traj

FEASIBLE_TRAJ_THRESHOLD = 1000

class MotionPrimitiveLibrary:
    def __init__(self, delta_yaw=21, delta_norm=15, tf=1):
        self.delta_yaw = delta_yaw
        self.delta_norm = delta_norm
        self.tf = tf
        self.trajs = []
        self.pos0, self.vel0, self.acc0 = np.zeros(3), np.zeros(3), np.zeros(3)
        self.goal_point, self.goal_direction = np.zeros(3), np.zeros(3)
        self.edt_function = None

    def set_init_state(self, pos0, vel0, acc0):
        self.pos0, self.vel0, self.acc0 = pos0, vel0, acc0
    
    def set_local_goal(self, goal_point, goal_direction):
        self.goal_point, self.goal_direction = goal_point, goal_direction

    def set_edt_function(self, edt):
        self.edt_function = edt

    def optimize(self):
        norm0 = np.linalg.norm(self.vel0)
        yaw0 = np.arctan2(self.vel0[1], self.vel0[0])

        # Optimize with the norm constant, along the yaw values
        # We use the Brent method to optimize
        f = self.build_cost_function(norm=norm0, z=self.pos0[2])
        yaw = scipy.optimize.brent(f, brack=(yaw0 - np.pi * .5, yaw0 + np.pi * .5), tol=1e-2)
        f = self.build_cost_function(yaw=yaw, z=self.pos0[2])
        norm = scipy.optimize.brent(f, brack=(np.clip(norm0 - 4/self.tf, .1, 1e5), norm0 + .5/self.tf), tol=1e-2)
        norm = min(norm, norm0 + .5/self.tf)
        
        traj = self.generate_traj(self.pos0, self.vel0, self.acc0, norm, yaw, self.pos0[2])
        traj.compute_cost(self.goal_point, self.goal_direction, self.edt_function)
        if traj._cost > FEASIBLE_TRAJ_THRESHOLD:
            raise TrajectoryError(traj)
        return traj


    # Generate a trajectory function cost function, by 2 of the scalar parameters
    def build_cost_function(self, norm=None, yaw=None, z=None):
        def cost(norm, yaw, z):
            traj = self.generate_traj(self.pos0, self.vel0, self.acc0, norm, yaw, z)
            traj.compute_cost(self.goal_point, self.goal_direction, self.edt_function)
            self.trajs.append(traj)
            return traj._cost

        if norm is not None and yaw is not None:
            def f(z):
                print('z=', z)
                return cost(norm, yaw, z)
        elif norm is not None:
            def f(yaw):
                print('yaw=', yaw)
                return cost(norm, yaw, z)
        else:
            def f(norm):
                print('norm=', norm)
                return cost(norm, yaw, z)
        return f

    def generate_traj(self, pos0, vel0, acc0, norm_velf, yawf, zf):
        traj = MotionPrimitive(pos0, vel0, acc0, [0, 0, -9.81])
        traj.set_goal_position([None, None, zf])
        traj.set_goal_velocity([norm_velf * np.cos(yawf), norm_velf * np.sin(yawf), 0])
        traj.set_goal_acceleration([0, 0, 0])
        traj.generate(self.tf)
        return traj
    

File: src/MotionPrimitiveLibrary.cpp
#include "MotionPrimitiveLibrary.h"

MotionPrimitiveLibrary::MotionPrimitiveLibrary(const double tf) : _tf(tf) {}

void MotionPrimitiveLibrary::setInitState(Vec3 pos0, Vec3 vel0, Vec3 acc0) {
    _pos0 = Vec3(pos0);
    _vel0 = Vec3(vel0);
    _acc0 = Vec3(acc0);
}

void MotionPrimitiveLibrary::setLocalGoal(Vec3 goalPoint, Vec3 goalDir) {
    _goalPoint = Vec3(goalPoint);
    _goalDir = Vec3(goalDir);
}

void MotionPrimitiveLibrary::setEDT(DynamicEDTOctomap* edt) {
    _edt = edt;
}

bool MotionPrimitiveLibrary::optimize() {
    using boost::math::tools::brent_find_minima;

    const double norm0 = _vel0.GetNorm2();
    const double yaw0 = atan2(_vel0.y, _vel0.x);

    std::cout << "*******************************" << std::endl;
    std::cout << "Optimizing..." << std::endl;
    std::cout << "\tInitial position " << _pos0 << std::endl;
    std::cout << "\tInitial velocity " << _vel0 << std::endl;
    std::cout << "\tInitial acceleration " << _acc0 << std::endl;
    std::cout << "\tInitial norm " << norm0 << std::endl;
    std::cout << "\tInitial yaw " << yaw0 << std::endl;

    // Optimize using the Brent's method, implemented in Boost
    // Optimize the yaw
    const yawCostFunc f1(_pos0, _vel0, _acc0, _goalPoint, _goalDir, norm0, _pos0.z, _tf, _edt, _trajs);
    std::pair<double, double> r = brent_find_minima(f1, yaw0 - PI * .5, yaw0 + PI * .5, 10);
    const double yaw = r.first;

    std::cout << "Yaw optimization result: " << r.first << std::endl;

    // Optimize the norm
    const normCostFunc f2(_pos0, _vel0, _acc0, _goalPoint, _goalDir, yaw, _pos0.z, _tf, _edt, _trajs);
    r = brent_find_minima(f2, std::max(norm0 - 4/_tf, .1), norm0 + .5/_tf, 10);
    const double norm = r.first;
    
    std::cout << "Norm optimization result: " << r.first << std::endl;
    
    // Optimize the z
    const double z = _pos0.z;

    std::cout << "Building the trajectory" << std::endl;
    _traj = buildTrajectory(_pos0, _vel0, _acc0, norm, yaw, z, _tf);
    if (_traj.GetCost(_goalPoint, _goalDir, _edt) > FEASIBLE_TRAJ_THRESHOLD) {
        return false;
    }
    return true;
}

MotionPrimitive MotionPrimitiveLibrary::buildTrajectory(
    const Vec3 pos0, const Vec3 vel0, const Vec3 acc0,
    const double normVelf, const double yawf, const double zf, const double tf
) {
    MotionPrimitive traj(pos0, vel0, acc0, Vec3(0, 0, -9.80665));
    traj.SetGoalPositionInAxis(2, zf);
    traj.SetGoalVelocity(Vec3(normVelf * cos(yawf), normVelf * sin(yawf), 0));
    traj.SetGoalAcceleration(Vec3(0, 0, 0));
    traj.Generate(tf);
    return traj;
}